### ISSUE TITLE
'File is a directory: undefined' if building with a symlinked directory present

### ISSUE THREAD

**Issue by danielunderwood** [resin-io/resin-cli#833]
- **resin-cli version:** 7.1.4
- **Operating system and architecture:**
    Workstation: Ubuntu 17.10 Artful, Linux 4.13.0-37 x86_64
    Target: RPi 3 on resinOS
- Docker version: 18.02.0-ce, build fc4de44
    

I'm running into the following error on both `resin build` and `resin deploy`

```
>>> resin build --application myapplication $DIR
[Info]    Creating default composition with source: /home/user@domain/Repos/myapplication
[Info]    Building for armv7hf/raspberrypi3
[Build]   Built 1 service in 0 seconds
[Build]   main Preparing...
[Error]   Build failed
File is a directory: undefined
```

Note that for `$DIR`, I've tried `.`, `$PWD`, leaving it off, and going to the directory above and doing `./myapplication/`. The contents of that directory are as follows:

```
Repos/myapplication
├── myapplication.egg-info (dir)
├── myapplication.py
├── myapplication.pyc
├── .cache (dir)
├── Dockerfile
├── .dockerignore
├── .git (dir)
├── git-requirements.txt
├── .idea (dir)
├── Jenkinsfile
├── __pycache__ (dir)
├── README.md
├── requirements.txt
├── service (dir)
├── setup.py
├── test_myapplication.py
├── .tox (dir)
├── tox.ini
├── venv (dir)
└── venv3 (dir)
```

I thought that it may be something with the service directory, but that didn't seem to help either.

__Edit:__ After a bit more tinkering, it seems that I can build by pushing to a resin git remote or on another Ubuntu 17.10 machine. The only difference with that machine is that I have installed npm via the repos instead of nvm and the home directories are `/home/domain/user` instead of `/home/user@domain`. I've seen other applications get tripped up on the `@` in the home path...maybe that's the issue?

<img src="https://frontapp.com/assets/img/icons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_c6hf)

**Comment by danielunderwood** [comment]
I ended up with this same error today, but on the other machine that was working. It was a Dockerfile issue in today's case as a plain `docker build` wouldn't work. I don't think the cause was the same since `docker build` did work on the Dockerfile when I reported the issue a couple days ago. Is there a way to turn on more verbose logging so we can see where docker hits an issue?

**Comment by pimterry** [comment]
@danielunderwood hmm, ok, I haven't seen this before. You can get slightly more verbose logging by setting the `DEBUG` env var to anything, e.g. `export DEBUG=1`. If it's possible, that should give you a stack trace.

Sounds like you've eliminated the `@` as a possible cause, since it's failing on another machine, is that right? I'm not sure exactly what's going on in the end here - are you saying any time you do a docker build which fails, you hit this odd error message?

There's some more background that might be useful - which npm/node/resin-cli version are you using, and are you using the standalone or npm build?

Thanks for the report, any info you can give me to help reproduce this will make it much easier to investigate further.

**Comment by danielunderwood** [comment]
I have resolved the issue on the other machine by correcting the Dockerfile template, but I'm still getting the error on the machine that had the issue in the first place, so it seems to me like they may be two separate issues. Though I'm not sure that then `@` is the problem since the logs below seem to be able to look into those directories.

Here are the current versions:
* npm 5.5.1
* node 8.9.1
* resin-cli 7.2.0 (updated from 7.1.4 in original issue)

After setting the debug var, I get the following output (note that I'm now using qemu emulation)
```
>>> resin deploy myapplication -e
[Debug]   Parsing input...
[Debug]   Loading project...
[Debug]   Resolving project...
[Debug]   Failed to resolve project: 
[Debug]   AggregateError of:
[Debug]       Error: ENOENT: no such file or directory, open '/home/user@domain/Repos/myapplication/resin-compose.yml'
[Debug]       Error: ENOENT: no such file or directory, open '/home/user@domain/Repos/myapplication/resin-compose.yaml'
[Debug]       Error: ENOENT: no such file or directory, open '/home/user@domain/Repos/myapplication/docker-compose.yml'
[Debug]       Error: ENOENT: no such file or directory, open '/home/user@domain/Repos/myapplication/docker-compose.yaml'
[Info]    Creating default composition with source: /home/user@domain/Repos/myapplication
[Debug]   Creating project...
[Info]    Building for rpi/raspberry-pi
[Info]    Emulation is enabled
[Build]   Built 1 service in 0 seconds
[Build]   main Preparing...
[Error]   Deploy failed
Error: EISDIR: illegal operation on a directory, read

If you need help, don't hesitate in contacting us at:

  GitHub: https://github.com/resin-io/resin-cli/issues/new
  Forums: https://forums.resin.io
```

**Comment by danielunderwood** [comment]
On a bit more inspection, I get different errors depending on if debugging is enabled or not:

```
jenkins@jenkins:/var/lib/jenkins/workspace/blep_resin-deploy-RNQ6KNWUFLAIZ5XXHB4VTDZQ3YR5D6GEMCSGGWY6DNP6OMTDWSNQ$ resin build --application blep
[Info]    Creating default composition with source: /var/lib/jenkins/workspace/blep_resin-deploy-RNQ6KNWUFLAIZ5XXHB4VTDZQ3YR5D6GEMCSGGWY6DNP6OMTDWSNQ
[Info]    Building for armv7hf/raspberrypi3
[Build]   Built 1 service in 0 seconds
[Build]   main Preparing...
[Error]   Build failed
File is a directory: undefined

If you need help, don't hesitate in contacting us at:

  GitHub: https://github.com/resin-io/resin-cli/issues/new
  Forums: https://forums.resin.io

jenkins@jenkins:/var/lib/jenkins/workspace/blep_resin-deploy-RNQ6KNWUFLAIZ5XXHB4VTDZQ3YR5D6GEMCSGGWY6DNP6OMTDWSNQ$ export DEBUG=1 && resin build --application blep
[Debug]   Parsing input...
[Debug]   Loading project...
[Debug]   Resolving project...
[Debug]   Failed to resolve project: 
[Debug]   AggregateError of:
[Debug]       Error: ENOENT: no such file or directory, open '/var/lib/jenkins/workspace/blep_resin-deploy-RNQ6KNWUFLAIZ5XXHB4VTDZQ3YR5D6GEMCSGGWY6DNP6OMTDWSNQ/resin-compose.yml'
[Debug]       Error: ENOENT: no such file or directory, open '/var/lib/jenkins/workspace/blep_resin-deploy-RNQ6KNWUFLAIZ5XXHB4VTDZQ3YR5D6GEMCSGGWY6DNP6OMTDWSNQ/resin-compose.yaml'
[Debug]       Error: ENOENT: no such file or directory, open '/var/lib/jenkins/workspace/blep_resin-deploy-RNQ6KNWUFLAIZ5XXHB4VTDZQ3YR5D6GEMCSGGWY6DNP6OMTDWSNQ/docker-compose.yaml'
[Debug]       Error: ENOENT: no such file or directory, open '/var/lib/jenkins/workspace/blep_resin-deploy-RNQ6KNWUFLAIZ5XXHB4VTDZQ3YR5D6GEMCSGGWY6DNP6OMTDWSNQ/docker-compose.yml'
[Info]    Creating default composition with source: /var/lib/jenkins/workspace/blep_resin-deploy-RNQ6KNWUFLAIZ5XXHB4VTDZQ3YR5D6GEMCSGGWY6DNP6OMTDWSNQ
[Debug]   Creating project...
[Info]    Building for armv7hf/raspberrypi3
[Build]   Built 1 service in 0 seconds
[Build]   main Preparing...
[Error]   Build failed
Error: EISDIR: illegal operation on a directory, read

If you need help, don't hesitate in contacting us at:

  GitHub: https://github.com/resin-io/resin-cli/issues/new
  Forums: https://forums.resin.io
```

It's a bit strange that this same user can clone the repo and proceed with the build fine, but the automated build seems to fail. I thought it may have been the long directory name, but sticking everything in `/tmp/application` seemed to fail with the same error, as did sticking it in the home directory (which is where a manual clone/build works).

**Comment by pimterry** [comment]
Hmm, interesting. I still can't reproduce this I'm afraid. I'll keep investigating next week, if you manage to come up with any clues as to what's different between the successful/failing cases though, that'd be very useful!

**Comment by emdash** [comment]
I just hit this myself. I am wondering if it is because I must run resin-cli as root, and I have an encrypted homedir.

**Comment by emdash** [comment]
For me the solution (temporarily) is to move my workspace for docker-compose to a global location under `/tmp`. In my case I'm pretty sure it's down to the docker daemon to being able to see my encrypted homedir. Not sure what it might be in your case @danielunderwood, but it couldn't hurt to try.

**Comment by emdash** [comment]
For me the solution (temporarily) is to move my workspace for docker-compose to a global location under `/`. In my case I'm pretty sure it's down to the docker daemon being unable to see my encrypted homedir. Not sure what it might be in your case @danielunderwood, but it couldn't hurt to try something similar.

**Comment by pimterry** [comment]
@danielunderwood @emdash if you try to do an equivalent build just using vanilla Docker directly with some small non-resin codebase, do you see the same issues?

If so, that's pretty strong evidence that there is some problem with your docker daemon and your system configuration somewhere. If not, then there's something odd in the way that the CLI is orchestrating docker here, and we'll need to work out what that is.

**Comment by danielunderwood** [comment]
Normal docker builds are working fine on both workstation and CI server that I was having this issue with.

I can't speak for the original error that I was having, but yesterday I was having the error again with a different Dockerfile. I tried to do a manual build of the image and ended up getting an error on one of my `apt-get install` commands. Perhaps this error is seen whenever there's an error in the docker build and it's not properly logging the docker error? It would be nice if we could grab the docker build log. I see that that option is available when using the resin git remote way of doing things, but I'm using the cli due to the need to pull in other private repos.

I don't think the docker error is the same thing happening on the CI build since I can build the same image manually, but it does appear to result in the same error message (even with the debug option). Maybe it's another case of the cli reporting an error, but not the underlying cause of the error?

**Comment by pimterry** [comment]
@danielunderwood I still can't see a clear cause for this. I think we should be able to get more info with proper trace of system calls though. At least knowing which file it's trying to read would be a good clue.

Can you run `strace -o trace.txt <your CLI command>` with a minimal reproduction that fails, and then send me the `trace.txt`? You can reach me at tim @ resin.io.

The trace might be too big to email, but it should zip pretty well, or you can host it somewhere else and send me a link instead if you prefer.

**Comment by Jaeyoung-Lim** [comment]
I have the same problem, any update on this?

**Comment by thgreasi** [comment]
Hey @Jaeyoung-Lim we are still investigating this.
It could be helpful if you could provide a stacktrace (as Tim mentions above) to help us further debug this.

**Comment by thisisbrians** [comment]
I'm experiencing the same issue on MacOS and was able to resolve it by copying my entire project directory into `/tmp` prior to building, per @emdash's suggestion.

**Comment by pimterry** [comment]
@thisisbrians I'd love to resolve this properly, but we still can't reproduce it, so we need a trace of the underlying OS operations from the CLI to tell where it's failing.

I think on MacOS strace doesn't exist, but a quick google suggests dtruss can do the same thing. Can you try running `sudo dtruss <your resin build command>` in the failing directory, and then send me the output at tim @ resin.io? That should give us a log of every system call the CLI makes, which will hopefully include a failing  call to read the specific mysterious directory that's causing problems here. It might be quite large, but it should zip pretty well, or you can host it somewhere and send me a link instead.

**Comment by bremac** [comment]
I just ran into this problem too. resin-cli was trying to open a directory named `lib64` inside a Python virtualenv. Here's the relevant snippet of my strace:
```
[pid 25325] openat(AT_FDCWD, "/home/bremac/factorytx/venv/lib64", O_RDONLY|O_CLOEXEC <unfinished ...>
[pid 25325] <... openat resumed> )      = 211
... much later, in a subprocess ...
[pid 25326] read(211,  <unfinished ...>
[pid 25326] <... read resumed> 0x557109d9b570, 8192) = -1 EISDIR (Is a directory)
```
I confirmed that moving the venv directory out of `/home/bremac/factorytx/` fixed the problem.

What's interesting is that the lib64 directory is a symlink to another directory:
```
$ ls -l venv/
total 8
drwxr-xr-x 3 bremac bremac 440 Apr 22 22:39 bin
drwxr-xr-x 3 bremac bremac  60 Apr 22 22:38 doc
drwxr-xr-x 2 bremac bremac  40 Apr 22 22:37 include
drwxr-xr-x 3 bremac bremac  60 Apr 22 22:37 lib
lrwxrwxrwx 1 bremac bremac   3 Apr 22 22:37 lib64 -> lib
-rw-r--r-- 1 bremac bremac  61 Apr 22 22:37 pip-selfcheck.json
-rw-r--r-- 1 bremac bremac  69 Apr 22 22:37 pyvenv.cfg
```

Presumably somewhere there's a logic error that causes all symlinks to be treated as a regular file.

**Comment by pimterry** [comment]
@bremac perfect, thank you! That makes sense, and I can now reproduce this locally. I'll get a fix put together asap.

**Comment by pimterry** [comment]
Resolved in Resin CLI 7.8.3
